src/services/XmppClient.js

```diff
@@ -224,6 +230,24 @@ class XmppClient {
       xmppClient.callbacks.chat.forEach((callback) => callback(message))
     }
 
+    // check message fasten (XEP-0422)
+    const fasten = stanza.getChildrenByFilter(child => child.attrs.xmlns === NS.MESSAGE_FASTENING)
+    if (fasten.length > 0) {
+      // check MUC retracted message (XEP-0425)
+      const moderation = fasten[0].getChildrenByFilter(child => child.attrs.xmlns === NS.MESSAGE_MODERATION)
+      if (moderation.length > 0) {
+        const retract = moderation[0].getChildrenByFilter(child => child.attrs.xmlns === NS.MESSAGE_RETRACTED)
+        if (retract.length > 0) {
+          const retracted = {
+            stanzaId: fasten[0].attrs.id,
+            reason: moderation[0].getChild('reason').getText(),
+            by: this.parseJid(moderation[0].attrs.by),
+          }
+          xmppClient.callbacks.messageRetracted.forEach((callback) => callback(retracted))
```

Comment:
I think we want to check that either the MUC supports the feature here, or that the user doing the retraction has the appropriate affiliation.

Reply:
As we just listen for this kind of messages, I do not understand what you want to check?
I mean, if the MUC does not support it, we would not receive fasten/moderation?

Reply:
Anybody can forge such a payload, it's just a regular message.

I noticed the spec says the message is to be sent by the room though, so maybe we can check this instead of the affiliation, or the capability.

Reply:
OK I'll add a check on the sender to keep only `groupchat` emitted by the muc